File: ios/Classes/Scanner Core/SingleFrameCardScanner.swift
//
//  SingleFrameCardScanner.swift
//  card_scanner
//
//  Created by Abhinav Kumar on 11/02/21.
//

import Foundation
import MLKitTextRecognition

class SingleFrameCardScanner {
    var cardScanOptions: CardScannerOptions
    
    init(withOptions cardScanOptions: CardScannerOptions) {
        self.cardScanOptions = cardScanOptions
    }
    
    func scanSingleFrame(visionText: Text) -> CardDetails? {
        guard let optionalCardNumberResult = CardNumberFilter(
            visionText: visionText,
            scannerOptions: cardScanOptions
        ).filter() as? CardNumberScanResult? else {
            debugLog("No card number found yet", scannerOptions: cardScanOptions)
            return nil
        }
        
        guard let cardNumberResult = optionalCardNumberResult else {
            return nil
        }
        
        guard cardNumberResult.cardNumber.isNotEmpty else {
            return nil
        }
        
        let cardExpiryResult: ExpiryDateScanResult? = ExpiryDateFilter(
            visionText: visionText,
            scannerOptions: cardScanOptions,
            cardNumberScanResult: cardNumberResult
        ).filter() as? ExpiryDateScanResult
        
        let cardHolderResult: CardHolderNameScanResult? = CardHolderNameFilter(
            visionText: visionText,
            scannerOptions: cardScanOptions,
            cardNumberScanResult: cardNumberResult
        ).filter() as? CardHolderNameScanResult
        
        return CardDetails(
            cardNumber: cardNumberResult.cardNumber,
            cardHolderName: cardHolderResult?.cardHolderName ?? "",
            expiryDate: cardExpiryResult?.expiryDate ?? ""
        )
    }
}

File: android/src/main/java/com/nateshmbhat/card_scanner/SingleFrameCardScanner.kt
package com.nateshmbhat.card_scanner

import com.google.mlkit.vision.text.Text
import com.nateshmbhat.card_scanner.scanner_core.models.CardDetails
import com.nateshmbhat.card_scanner.scanner_core.models.CardScannerOptions
import com.nateshmbhat.card_scanner.scanner_core.scan_filters.CardHolderNameFilter
import com.nateshmbhat.card_scanner.scanner_core.scan_filters.CardNumberFilter
import com.nateshmbhat.card_scanner.scanner_core.scan_filters.ExpiryDateFilter

class SingleFrameCardScanner(private val scannerOptions: CardScannerOptions) {
  fun scanSingleFrame(visionText: Text): CardDetails? {
    val cardNumberResult = CardNumberFilter(visionText, scannerOptions).filter();
    if (cardNumberResult?.cardNumber?.isEmpty() != false) {
      return null;
    }
    val cardExpiryResult = ExpiryDateFilter(visionText, scannerOptions, cardNumberResult).filter();
    val cardHolderResult = CardHolderNameFilter(visionText, scannerOptions, cardNumberResult).filter();
    return CardDetails(cardNumber = cardNumberResult.cardNumber, expiryDate = cardExpiryResult?.expiryDate
            ?: "", cardHolderName = cardHolderResult?.cardHolderName ?: "");
  }
}